pkg/registry/servicecatalog/clusterid/storage.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterid
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/generic"
+	"k8s.io/apiserver/pkg/registry/generic/registry"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage"
+
+	scmeta "github.com/kubernetes-incubator/service-catalog/pkg/api/meta"
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog"
+	"github.com/kubernetes-incubator/service-catalog/pkg/registry/servicecatalog/server"
+)
+
+// NewStorage creates a new rest.Storage responsible for accessing ServiceInstance
+// resources
+func NewStorage(opts server.Options) rest.Storage {
+	prefix := "/" + opts.ResourcePrefix()
+
+	storageInterface, dFunc := opts.GetStorage(
+		&servicecatalog.ClusterID{},
+		prefix,
+		clusterIDStrategies,
+		NewList,
+		nil,
+		storage.NoTriggerPublisher,
+	)
+
+	store := registry.Store{
+		NewFunc:     EmptyObject,
+		NewListFunc: NewList,
+		KeyRootFunc: opts.KeyRootFunc(),
+		KeyFunc:     opts.KeyFunc(false),
+		// Retrieve the name field of the resource.
+		ObjectNameFunc: func(obj runtime.Object) (string, error) {
+			return scmeta.GetAccessor().Name(obj)
+		},
+		// Used to match objects based on labels/fields for list.
+		PredicateFunc: nil,
+		// DefaultQualifiedResource should always be plural
+		// TODO: there can be only 1.
+		DefaultQualifiedResource: servicecatalog.Resource("clusterids"),
+
+		CreateStrategy:          clusterIDStrategies,
+		UpdateStrategy:          clusterIDStrategies,
+		DeleteStrategy:          clusterIDStrategies,
+		EnableGarbageCollection: false,
+
+		Storage:     storageInterface,
+		DestroyFunc: dFunc,
+	}
+
+	options := &generic.StoreOptions{
+		RESTOptions: opts.EtcdOptions.RESTOptions,
+		AttrFunc:    nil,
+	}
+	if err := store.CompleteWithOptions(options); err != nil {
+		panic(err) // TODO: Propagate error up
+	}
+
+	return &store
+}
+
+// NewList returns a new shell of a ID list
+// TODO: I don't ever want to return a list.
```

Comment:
I think we do want to be able to return a list. Even though we want there to be at most a single `ClusterID`, we still want to support a kubectl query of all the `ClusterID` resources.